midend/flattenLogMsg.cpp

```diff
@@ -0,0 +1,198 @@
+#include "midend/flattenLogMsg.h"
+#include "flattenLogMsg.h"
+
+namespace P4 {
+
+const IR::Type_StructLike* FindTypeInLogMsgForReplace::hasStructInParameter(
+        const IR::MethodCallStatement* methodCallStatement) {
+    if (auto* path = methodCallStatement->methodCall->method->to<IR::PathExpression>()) {
+        if (path->path->name.name == "log_msg") {
+            auto* param1 = methodCallStatement->methodCall->arguments->at(0)->expression;
+            auto* param2 = methodCallStatement->methodCall->arguments->at(1)->expression;
+            if (!param1->is<IR::StringLiteral>() || !param2->is<IR::StructExpression>()) {
+                return nullptr;
+            }
+            auto* type =
+                typeMap->getTypeType(methodCallStatement->methodCall->typeArguments->at(0), true);
+            if (auto* typeStruct = type->to<IR::Type_StructLike>()) {
+                for (auto field : typeStruct->fields) {
+                    if (field->type->is<IR::Type_StructLike>()) {
+                        return typeStruct;
+                    }
+                }
+            }
+        }
+    }
+    return nullptr;
+}
+
+const IR::MethodCallStatement* FindTypeInLogMsgForReplace::prepareLogMsgStatement(
+        const IR::MethodCallStatement* methodCallStatement) {
+    if (auto* newMethod = getReplacementMethodCall(methodCallStatement->id)) {
+        return newMethod;
+    }
+    // Create new statement
+    auto* param1 = methodCallStatement->methodCall->arguments->at(0)->expression;
+    std::string strParam1 = param1->to<IR::StringLiteral>()->value.c_str();
+    auto* argType =
+        typeMap->getTypeType(methodCallStatement->methodCall->typeArguments->at(0), true);
+    auto* structType = argType->to<IR::Type_StructLike>();
+    index = 0;
+    auto exprVector =
+        unfoldStruct(methodCallStatement->methodCall->arguments->at(1)->expression, strParam1);
+    structType = generateNewStructType(structType, exprVector.first);
+    auto* newMethodCall = methodCallStatement->methodCall->clone();
```

Comment:
alternatively you can let the type inference compute this argument after clearTypeMap, by invoking it with a suitable 'true' flag to infer missing types.

Reply:
This type is used for creation of IR::StructExpression. The IR::StructExpression haven't  any constructors without IR::Type. Do I need to use IR::Type_Unknown instend of structType?

Reply:
Then you need to create the type, there isn't enough information for the typechecked to infer the type from the argument.